src/gennylib/include/gennylib/topology.hpp

```diff
@@ -0,0 +1,219 @@
+            mongoses[i].accept(v);
+            v.visitMongosesBetween(*this);
+        }
+        mongoses[mongoses.size() - 1].accept(v);
+        v.visitMongosesPost(*this);
+
+        v.visitShardedDescriptionPost(*this); 
+    }
+};
+
+/**
+ * Represents the topology of a MongoDB cluster.
+ */
+class Topology {
+public:
+    Topology(mongocxx::client& client) : _factory{client.uri().to_string()} {
+        MongoService service(client.uri().to_string());
+        update(service);
+    }
+
+    Topology(DBService& service) : _factory{service.uri()} {
+        update(service);
+    }
+
+    /**
+     * Traverse the cluster, using the visitor to act on it.
+     */
+    void accept(TopologyVisitor& v) { 
+        if (_topology) {
+            v.visitTopLevelPre(*_topology);
+            _topology->accept(v); 
+            v.visitTopLevelPost(*_topology);
+        }
+    }
+
+    /**
+     * Update the Topology's view of the cluster.
+     */
+    void update(DBService& service);
+
+private:
+    v1::PoolFactory _factory;
+    std::string nameToUri(const std::string& name);
+    void getDataMemberConnectionStrings(DBService& service);
+    void findConnectedNodesViaMongos(DBService& service);
+    std::unique_ptr<AbstractTopologyDescription> _topology = nullptr;
+};
+
+class ToJsonVisitor : public TopologyVisitor {
+public:
+    void visitTopLevelPre(const AbstractTopologyDescription&) {
+        result.str("");
+        result.clear();
+    }
+
+    void visitMongodDescription(const MongodDescription& desc) { result << "{mongodUri: " << desc.mongodUri << "}"; }
+    void visitMongodDescriptionBetween(const ReplSetDescription& desc) { result << ", "; }
+
+    void visitMongosDescription(const MongosDescription& desc) { result << "{mongosUri: " << desc.mongosUri << "}"; }
+    void visitMongosDescriptionBetween(const ReplSetDescription& desc) { result << ", "; }
+
+    void visitReplSetDescriptionPre(const ReplSetDescription& desc) {
+        if (desc.configsvr) {
+            result << "configsvr: ";
+        }
+        result << "{primaryUri: " << desc.primaryUri << ", nodes: [";
+    }
+    void visitReplSetDescriptionPost(const ReplSetDescription& desc) { result << "]}"; }
+
+    void visitShardedDescriptionPre(const ShardedDescription&) { result << "{"; }
+    void visitShardedDescriptionPost(const ShardedDescription&) { result << "}"; }
+
+    void visitShardsPre(const ShardedDescription&) { result << " shards: ["; }
+    void visitShardsBetween(const ShardedDescription&) { result << ", "; }
+    void visitShardsPost(const ShardedDescription&) { result << "], "; }
+
+    void visitMongosesPre(const ShardedDescription&) { result << " mongoses: ["; }
+    void visitMongosesBetween(const ShardedDescription&) { result << ", "; }
+    void visitMongosesPost(const ShardedDescription&) { result << "]"; }
+
+    std::string str() { return result.str(); }
+
+private:
+    std::stringstream result;
+};
```

Comment:
impls can be in .cpp (I realize it's annoying so my apologies for asking for banal code ceremony)